src/mlpack/core/data/save_impl.hpp

```diff
@@ -197,6 +197,45 @@ bool Save(const std::string& filename,
   return true;
 }
 
+// Save a Sparse Matrix
+template<typename eT>
+bool Save(const std::string& filename,
+          const arma::SpMat<eT>& matrix,
+          const bool fatal,
+          bool transpose)
+{
+  Timer::Start("saving_data");
+  arma::file_type saveType = arma::coord_ascii;
```

Comment:
There are multiple types of sparse formats that can be loaded; I'd suggest that we use `Extension()` like for the dense matrices, and if the extension is not `.txt`, we don't infer that the type is `coord_ascii`.  Also, can we use `csv_ascii` when the type is `.csv`, and `arma_binary` when the type is `.bin`?

Reply:
If I am correct,
The above suggestion would help answer this question: (raised in the description of the PR)

> What if an arma::mat contains a sparse matrix ? How to load that?
> Answer: If a sparse matrix is in dense format (in a csv or tsv having a lot of zeros), The **_user_** can load it using arma::mat(regular mlpack data::Load()) and then convert to arma::sp_mat

So please clarify this doubt:
input files for arma::sp_mat can be of 2 types:
1. Dense representation (having lots of zeros) which would be loaded with the help of arma::csv_ascii, arma::arma_ascii, arma::arma_binary etc.

2. arma::coord_ascii (of shape [n, 3]) which would load the sparse matrix using locations and values.

Are these types mutually exclusive ?
Or do I have to write some logic to determine either of the format types ?

Here, I understand the need of #2084 where manual type specification would have helped much. The user would simply specify arma::coord_ascii in the function call to load the file. 

Could you provide clarity on this ?

Reply:
Right, so there is some ambiguity in the CSV format.  The same Armadillo code, loading a sparse matrix with type `csv_ascii`, work for both of these CSV files:

("narrow" 3-column format)
```
1, 2, 3
1, 5, 6
7, 10, 11
```

("wide" dense format)
```
1, 2, 3, 4, 5
6, 7, 8, 0, 10
10, 0, 0, 6, 4
```

So, Armadillo does this format detection internally.  I think that we can just hand the file to Armadillo and let it do what it automatically does, using the file extension to auto-detect the format (`.csv` -> `csv_ascii`, `.txt` -> `coord_ascii`, `.bin` -> `arma_binary`).  If the format detection doesn't work correctly, then I think the user should specify a different file extension / file type, or try to load it as a dense matrix.  The user does have control over whether the matrix is loaded as dense or sparse simply via the type of the matrix they are loading.

Let me know what you think.  It's possible that I've overlooked something.